### PR TITLE
Unify generics code

### LINKED ISSUE #290: Unify generics code
**Is your feature request related to a problem? Please describe.**
The code for iterating over sequences and collections is scattered across multiple classes due to the fact that golang has no support for generics.

**Describe the solution you'd like**
Write a generic iterator code once and then wrappers to perform appropriate type coercion . Possible candidates are:

- [ ] `Iterator` and  derived types
- [ ] `FiniteIterator` and derived types

**Describe alternatives you have considered**
[Golang does not support generics](https://appliedgo.net/generics/) .

### PR DESCRIPTION
**Pull request by hsequeda**
Fixes #290 

**Changes:**
-Added Iterator interface.
-Implementing Iterator interface in util packege with GenericIterator type.
-Replaced all old iterators implementation for GenericIterator.
-Added test and comments for GenericIterator type.

**Does this change need to mentioned in CHANGELOG.md?**
 no

**Requires testing**
Requires testing
 no

**Comments about testing , should you have some**

### CODE REVIEW

**Comment by olemis** [comment]
@stdevHsequeda good , will take a look into this shortly

File: src/core/generic.go [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package core
+
+// Iterator defines a generic iterator.
+type Iterator interface {
+	CurrentData(interface{}) error
```
Comment (olemis) [inline src/core/generic.go:5]: This needs be changed

**Review by olemis — CHANGES_REQUESTED** [review]

File: src/core/generic.go (hunk 1 above)
Comment (hsequeda) [reply src/core/generic.go:5]: To Value() interface?

Comment (olemis) [reply src/core/generic.go:5]: yep

**Comment by olemis** [comment]
I'm starting to hesitate about whether moving forward with this is a good idea

**Comment by olemis** [comment]
I am not satisfied with the approaches I've seen so far ... Let's keep things the way they are now and reopen this should situation change at some point later

**Pull request closed without merging**